### PR TITLE
feat (dashboard): support docker image pull secrets for run

### LINKED ISSUE #2855: dashboard: support for docker pull secrets in run

### PR DESCRIPTION
**Pull request by dbm03**
### **User description**
Resolves #2855

___

### **PR Type**
Enhancement

___

### **Description**
- Added support for Docker image pull secrets in Run services
- Implemented new input field for pull credentials in the ServiceForm component
- Updated GraphQL queries and types to include the new pullCredentials field
- Added a link to documentation about using private registries
- Updated ServicesList component to handle pull credentials
- Added changeset for the new feature
- Made minor styling adjustments throughout the affected components

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>ServiceForm.tsx</strong><dd><code>Add pull credentials input to ServiceForm</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

dashboard/src/features/orgs/projects/services/components/ServiceForm/ServiceForm.tsx

<li>Added input field for pull credentials<br> <li> Added link to documentation about private registries<br> <li> Updated form submission to include pull credentials<br> <li> Minor styling adjustments<br>

</details>

  </td>
  <td><a href="https://github.com/nhost/nhost/pull/2906/files#diff-a02746694d45a84390d09b49a1b3eec85c25a8bd9a70b4834ee5af1ba82cb88e">+52/-5</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>ServiceFormTypes.ts</strong><dd><code>Update ServiceFormTypes with pull credentials</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

dashboard/src/features/orgs/projects/services/components/ServiceForm/ServiceFormTypes.ts

- Added pullCredentials to validation schema

</details>

  </td>
  <td><a href="https://github.com/nhost/nhost/pull/2906/files#diff-e9e0545b8c213ce04a08f9b04aedc81f96a031429e2ac9ac9e19d47982c112dc">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>ServicesList.tsx</strong><dd><code>Update ServicesList to include pull credentials</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

dashboard/src/features/orgs/projects/services/components/ServicesList/ServicesList.tsx

<li>Added pullCredentials to service form initial data<br> <li> Minor styling adjustments<br>

</details>

  </td>
  <td><a href="https://github.com/nhost/nhost/pull/2906/files#diff-643c818a248c42950336289392ac97ed9ef5c670ff8e47b80588b9802844d28a">+9/-8</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>graphql.ts</strong><dd><code>Update GraphQL types for pull credentials</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

dashboard/src/utils/__generated__/graphql.ts

- Updated GraphQL types to include pullCredentials field

</details>

  </td>
  <td><a href="https://github.com/nhost/nhost/pull/2906/files#diff-fbd5db84b560b1c91675004448c6c7fa0dcbfb28b9eb05d53b03e6cb7b83ebac">+6/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>getRunService.graphql</strong><dd><code>Update getRunService GraphQL query</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

dashboard/src/gql/services/getRunService.graphql

- Added pullCredentials field to getRunService query

</details>

  </td>
  <td><a href="https://github.com/nhost/nhost/pull/2906/files#diff-2e51dd3f3ee7a2bafdb9c754e57ce461cdf2b2ef754edc7d99120fa859e6c1d2">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>getRunServices.graphql</strong><dd><code>Update getRunServices GraphQL query</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

dashboard/src/gql/services/getRunServices.graphql

- Added pullCredentials field to RunServiceConfig fragment

</details>

  </td>
  <td><a href="https://github.com/nhost/nhost/pull/2906/files#diff-bddccfbe4a76e754eff4aff107d8780424c0f3606c781a8e6e3a5c9c099b7c57">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>soft-goats-juggle.md</strong><dd><code>Add changeset for pull credentials feature</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.changeset/soft-goats-juggle.md

- Added changeset for new pull credentials feature

</details>

  </td>
  <td><a href="https://github.com/nhost/nhost/pull/2906/files#diff-7a3a661b844b08a2582369644615cb42fd06c46aeb8d284a802faa2f1afbb140">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: dashboard/src/features/orgs/projects/services/components/ServiceForm/components/ImageFormSection/ImageFormSection.tsx [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+              >
+                <InfoIcon
+                  aria-label="Info"
+                  className="h-4 w-4"
+                  color="primary"
+                />
+              </Tooltip>
+            </Box>
+          }
+          placeholder="Add the secret variable for pull credentials here. e.g. {{ secrets.CONTAINER_REGISTRY_CREDENTIALS }}"
```
Comment (dbarrosop) [inline dashboard/src/features/orgs/projects/services/components/ServiceForm/components/ImageFormSection/ImageFormSection.tsx:146]:
```suggestion
          placeholder="Enter your for pull credentials here."
```

File: dashboard/src/features/orgs/projects/services/components/ServiceForm/components/ImageFormSection/ImageFormSection.tsx [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+          <Text>
+            Learn more about{' '}
+            <Link
+              href="https://docs.nhost.io/guides/run/registry#using-your-own-private-registry"
+              target="_blank"
+              rel="noopener noreferrer"
+              underline="hover"
+              className="font-medium"
+            >
+              Using your own private registry for images
```
Comment (dbarrosop) [inline dashboard/src/features/orgs/projects/services/components/ServiceForm/components/ImageFormSection/ImageFormSection.tsx:166]:
```suggestion
              using your own private registry for images
```

File: dashboard/src/features/orgs/projects/services/components/ServiceForm/components/ImageFormSection/ImageFormSection.tsx [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+  return (
+    <Box className="space-y-4 rounded border-1 p-4">
+      <Box className="flex flex-row items-center space-x-2">
+        <Text variant="h4" className="font-semibold">
+          Image
+        </Text>
+        <Tooltip
+          title={
+            <span>
+              Image to use, it can be hosted on any public registry or it can
```
Comment (dbarrosop) [inline dashboard/src/features/orgs/projects/services/components/ServiceForm/components/ImageFormSection/ImageFormSection.tsx:47]: This text seems to be a bit outdated, maybe the tooltip in image can just be removed

**Comment by dbarrosop** [comment]
Can we ensure the value isn't lost when switching types? I.e. if you have a public image and you click in Private registry and then switch back to public image, the value is lost, it'd be nice if it could be saved. At least the value that came from the configuration.

File: dashboard/src/features/orgs/projects/services/components/ServiceForm/ServiceForm.tsx [hunk 4]
```diff
@@ -88,14 +84,53 @@ export default function ServiceForm({
...
+  );
+
+  const handleImageTypeChange = (value: 'public' | 'private' | 'nhost') => {
+    if (value === initialImageType) {
+      setValue('image', initialData?.image);
+      setValue('pullCredentials', initialData?.pullCredentials);
+    } else {
+      setValue('image', '');
+      setValue('pullCredentials', undefined);
+    }
```
Comment (onehassan) [inline dashboard/src/features/orgs/projects/services/components/ServiceForm/ServiceForm.tsx:129]: minor thing, I think it's better to use `reset({image: ..., pullCredentials: ....})` to update multiple values at once

**Review by onehassan — APPROVED** [review]
left a minor comment but LGTM overall

**Pull request merged**